### DISCUSSION TITLE
Unix socket bidirectional communication

### DISCUSSION THREAD

**Discussion by caramboleyo** [denoland/deno#15363] [Q&A]
With Deno.listen i can listen to a unix socket file and with Deno.connect i can send data to that socket from another app. What does not work is to have it bidirectional. When i listen to a socket from another app, it deletes and recreates the socket file, the first app listening to it no longer reacts when sending. In https://github.com/denoland/deno/issues/9052 @bnoordhuis says this is expected behavior and that node does it the same way, but i had this bidirectional communication with node a few years ago. I was playing around the whole day (mostly to understand how the buffer works, this worked out of the box with node) and this is what I came up with:
(I also tried [Alice/Bob example](https://github.com/denoland/deno/blob/main/cli/tests/unit/net_test.ts#L465) with Deno.listenDatagram but that does not work when you put alice and bob in separate scripts).

service.js:
```javascript
async function listenToUnixSocket(listener) {
	for await (const connection of listener) {

		// receive binary chunks
		const chunkSize = 256;
		let totalSize = 0;
		let chunks = [];
		let chunkBuffer = new Uint8Array(chunkSize);
		let current;
		while (current = await connection.read(chunkBuffer)) {
			totalSize += current;
			chunks.push(chunkBuffer);
			chunkBuffer = new Uint8Array(chunkSize);
		}

		// merge chunks
		let mergedBuffer = new Uint8Array(totalSize);
		let offset = 0;
		for (let chunk of chunks) {
			if (totalSize < chunkSize) {
				chunk = chunk.slice(0, totalSize);
			}
			mergedBuffer.set(chunk, offset);
			offset += chunk.byteLength;
		}

		console.log('cli>', new TextDecoder().decode(mergedBuffer));
		await connection.write(new TextEncoder().encode('Response111')); //<-broken pipe because cli not listening
	}
}
const listener = Deno.listen({ path: 'cli.sock', transport: 'unix' });
listenToUnixSocket(listener);

```
cli.js:
```js
#!/usr/bin/env -S deno run -A --unstable --allow-env
const conn = await Deno.connect({ path: 'cli.sock', transport: "unix" });
conn.write(new TextEncoder().encode(Deno.args));
// how to listen here?
```

I dont understand why you delete the socket file. The main reason for unix sockets is for different applications to communicate with each other, even if they use completly different programming languages. Or did i get something wrong about that?

**Comment by bnoordhuis** [comment]
> `// how to listen here?`

Wrong question. The client doesn't listen, it reads.

**Reply by caramboleyo** [reply]
Yes, of corse. I tried read, it does not have that method when you get the connection from Deno.connect.
I thought i have to listen so the stream stays open, but the cli script sends the message, then exits immediately.

**Comment by kitsonk** [comment]
There are several abstractions in `deno.land/x` that provide this in a more user friendly way:

- https://deno.land/x/tcp_socket@0.0.2
- https://deno.land/x/socket_fetch@v0.1
- https://deno.land/x/socks@0.0.3

While a lot of libraries say "listen" that is a higher order abstraction, as @bnoordhuis, you are reading and writing to the socket.  A decent example is in socks [/lib/client.ts](https://deno.land/x/socks@0.0.3/lib/client.ts?code).  A server then is simply an abstraction on top of a client.

**Reply by caramboleyo** [reply]
All three links are tcp sockets. To me tcp is an internet protocol. I dont need network at all. I directly read/write the socket file (like i am used to inside linux machines). The client.ts looks nice, i see it gets a connection in the constructor. I will play around with that.

**Comment by alexindigo** [comment]
I'm seeing similar issue, that `Deno.listen(<file>, {transport: "unix"})` creates new socket file, and if file already exists it throws, not sure what's the idea behind it, and when I try to connect multiple "clients" (via Deno.connect), it throws too, while the same code with single client, it kind of works.

Also, even with single client, when I do `connection.close()` after I wrote into that socket, both client and server throw.
And there is not example in the docs, for how unix socket supposed to behave.
